### PR TITLE
feat: create SVG python bindings for new detector

### PR DESCRIPTION
**Pull request by asalzburger**
This PR introduces python bindings for the new detector and gives convenience functions to select volumes, restrict views and color them eventually.

An example is show cased in `detector_creation` which would need `thirdparty/OpenDataDetector` with version `v3.0.3`.

This would produce things like this:

Screenshot 2024-01-19 at 16 45 40

Screenshot 2024-01-19 at 16 45 56

### CODE REVIEW

File: Core/src/Geometry/Extent.cpp [hunk 1]
```diff
@@ -104,6 +104,17 @@ void Acts::Extent::setEnvelope(const ExtentEnvelope& envelope) {
   m_envelope = envelope;
 }
 
+bool Acts::Extent::contains(const Vector3& vtx) const {
+  Extent checkExtent;
+  for (int ibv = 0; ibv < (int)binValues; ++ibv) {
+    if (constrains((BinningValue)ibv)) {
```
Comment (niermann999) [inline Core/src/Geometry/Extent.cpp:110]: Is this a c-style cast to ```BinningValue```? Is is possible to do that with ```static_cast``` or smth like that?

**Review by niermann999 — COMMENTED** [review]
Looks good in general

File: Core/src/Geometry/Extent.cpp (hunk 1 above)
Comment (asalzburger) [reply Core/src/Geometry/Extent.cpp:110]: Ha ha, that was co-pilot writing it, you are right, let's static cast.

File: Examples/Python/src/Geometry.cpp [hunk 2]
```diff
@@ -183,13 +183,18 @@ void addExperimentalGeometry(Context& ctx) {
...
-  // Detector definition
-  py::class_<Detector, std::shared_ptr<Detector>>(m, "Detector");
-
   // Detector volume definition
   py::class_<DetectorVolume, std::shared_ptr<DetectorVolume>>(m,
                                                               "DetectorVolume");
 
+  // Detector definition
+  py::class_<Detector, std::shared_ptr<Detector>>(m, "Detector")
+      .def("number_volumes",
```
Comment (paulgessinger) [inline Examples/Python/src/Geometry.cpp:192]: Could be a property

File: Examples/Python/src/Svg.cpp [hunk 3]
```diff
@@ -30,12 +38,109 @@ using namespace pybind11::literals;
 using namespace Acts;
 using namespace ActsExamples;
 
+namespace {
+
+using ViewAndRange = std::tuple<std::string, Acts::Extent>;
+
+// Helper function to be picked in different access patterns
+static inline actsvg::svg::object viewDetectorVolume(
```
Comment (paulgessinger) [inline Examples/Python/src/Svg.cpp:46]:
```suggestion
actsvg::svg::object viewDetectorVolume(
```

File: Examples/Python/src/Svg.cpp [hunk 4]
```diff
@@ -30,12 +38,109 @@ using namespace pybind11::literals;
...
+    const ViewAndRange& viewAndRange) {
+  actsvg::svg::object svgDet;
+  svgDet._id = identification;
+  svgDet._tag = "g";
+
+  auto [view, viewRange] = viewAndRange;
+
+  // The surfaces to be drawn
+  std::vector<Svg::ProtoVolume::surface_type> surfaces;
+  // If there is view reange restriction, filter the surfaces
```
Comment (paulgessinger) [inline Examples/Python/src/Svg.cpp:57]:
```suggestion
  // If there is view range restriction, filter the surfaces
```

File: Examples/Python/src/Svg.cpp [hunk 5]
```diff
@@ -30,12 +38,109 @@ using namespace pybind11::literals;
...
+      svgDet.add_object(Svg::View::zr(vs, identification));
+    } else {
+      throw std::invalid_argument("Unknown view type");
+    }
+  }
+  return svgDet;
+};
+
+// Helper function to be picked in different access patterns
+static inline void viewDetector(
```
Comment (paulgessinger) [inline Examples/Python/src/Svg.cpp:89]:
```suggestion
void viewDetector(
```

File: Examples/Python/src/Svg.cpp [hunk 6]
```diff
@@ -30,12 +38,109 @@ using namespace pybind11::literals;
...
+}  // namespace
+
 namespace Acts::Python {
 void addSvg(Context& ctx) {
   auto [m, mex] = ctx.get("main", "examples");
 
+  auto svg = m.def_submodule("svg");
+
+  // Some basics
+  { py::class_<actsvg::svg::object>(svg, "object"); }
```
Comment (paulgessinger) [inline Examples/Python/src/Svg.cpp:139]:
```suggestion
  py::class_<actsvg::svg::object>(svg, "object");
```

File: Examples/Scripts/Python/detector_creation.py
Comment (paulgessinger) [inline Examples/Scripts/Python/detector_creation.py:1]: We should get this running in the CI ASAP. When can we bump ODD to get ready for this?

Comment (asalzburger) [reply Examples/Scripts/Python/detector_creation.py:1]: We can bump ODD to v3.0.3 practically immediately, the changes to v3 are marginal (but we will need to change all reference files).

Comment (paulgessinger) [reply Examples/Scripts/Python/detector_creation.py:1]: That's ok as long as it's done consistently.

File: Examples/Python/src/Svg.cpp (hunk 3 above)
Comment (asalzburger) [reply Examples/Python/src/Svg.cpp:46]: fixed.

File: Examples/Python/src/Geometry.cpp (hunk 2 above)
Comment (asalzburger) [reply Examples/Python/src/Geometry.cpp:192]: Don't know what this means ... a python property of the detector?

File: Examples/Python/src/Svg.cpp (hunk 3 above)
Comment (asalzburger) [reply Examples/Python/src/Svg.cpp:46]: Fixed.

File: Examples/Python/src/Svg.cpp (hunk 5 above)
Comment (asalzburger) [reply Examples/Python/src/Svg.cpp:89]: Fixed.

File: Examples/Python/src/Svg.cpp (hunk 6 above)
Comment (asalzburger) [reply Examples/Python/src/Svg.cpp:139]: Done.

File: Examples/Python/src/Svg.cpp (hunk 4 above)
Comment (asalzburger) [reply Examples/Python/src/Svg.cpp:57]: Fixed.

File: Examples/Scripts/Python/detector_creation.py
Comment (asalzburger) [reply Examples/Scripts/Python/detector_creation.py:1]: Let's try to do this during the week.

File: Core/src/Geometry/Extent.cpp (hunk 1 above)
Comment (niermann999) [reply Core/src/Geometry/Extent.cpp:110]: Does it not work?

Comment (niermann999) [inline Core/src/Geometry/Extent.cpp:110]: Does the static cast not work?

Comment (asalzburger) [reply Core/src/Geometry/Extent.cpp:110]: It should, I just forgot - will fix it.

Comment (asalzburger) [reply Core/src/Geometry/Extent.cpp]: Changed consistently to loop through `s_binningValues` as we have it exactly for this reason.

**Review by niermann999 — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [21d6c53983997133c7cebe8cf8e91a23ec922aee]
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/6772769)!

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462810)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsBenchmarkWithSpot.sh 8 100](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462809)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462808)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateAmbiguityResolution.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462807)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateResolvedTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462806)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462805)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462804)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462803)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462802)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462801)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462800)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462799)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsGSFRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462798)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462797)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462796)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462795)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462794)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462793)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462792)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462791)||
|:red_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462790)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462789)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35462788)||

**Pull request merged**